### PR TITLE
Plfm 6161 id parser

### PR DESCRIPTION
**Pull request by john-hill**

### CODE REVIEW

File: lib/lib-schema-id/pom.xml [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+		<artifactId>lib</artifactId>
+		<version>develop-SNAPSHOT</version>
+	</parent>
+	<artifactId>lib-schema-id</artifactId>
+	<name>lib-schema-id</name>
+	<url>http://maven.apache.org</url>
+	<properties>
+		<project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+		<jacoco.branch.minumum>0.50</jacoco.branch.minumum>
+		<jacoco.line.minumum>0.50</jacoco.line.minumum>
```
Comment (marcomarasca) [inline lib/lib-schema-id/pom.xml:18]: You could potentially exclude some paths in the coverage, e.g. stuff autogenerated that does not need to be tested by us (See the root pom.xml to see how to exclude)

File: lib/lib-schema-id/src/test/java/org/sagebionetworks/schema/SchemaIdParserTest.java [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+	public void testVersionCoreMinorLeadingZeror() throws ParseException {
+		SchemaIdParser parser = new SchemaIdParser("1.05.2");
+		assertThrows(ParseException.class, ()->{
+			parser.versionCore();
+		});
+	}
+	
+	@Test
+	public void testVersionCorePatchLeadingZeror() throws ParseException {
+		SchemaIdParser parser = new SchemaIdParser("1.5.02");
```
Comment (marcomarasca) [inline lib/lib-schema-id/src/test/java/org/sagebionetworks/schema/SchemaIdParserTest.java:78]: This should probably throw

File: lib/lib-schema-id/src/test/java/org/sagebionetworks/schema/SchemaIdParserTest.java [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+		VersionCore core = parser.versionCore();
+		// confirm the 2 is lost
+		assertEquals("1.5.0", core.toString());
+	}
+	
+	@Test
+	public void testAlphaNumeric() throws ParseException {
+		testAlphanumericIdentifier("-");
+		testAlphanumericIdentifier("a");
+		testAlphanumericIdentifier("a1123");
```
Comment (marcomarasca) [inline lib/lib-schema-id/src/test/java/org/sagebionetworks/schema/SchemaIdParserTest.java:89]: Add case aaa123

Comment (marcomarasca) [reply lib/lib-schema-id/src/test/java/org/sagebionetworks/schema/SchemaIdParserTest.java:89]: Also single 0 is allowed

File: lib/lib-schema-id/src/test/java/org/sagebionetworks/schema/SchemaIdParserTest.java [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+		Build build = parser.build();
+		assertEquals(build.toString(), input);
+	}
+	
+	@Test
+	public void testSemanticVersion() throws ParseException {
+		testSemanticVersion("0.0.0");
+		testSemanticVersion("1.23.456");
+		testSemanticVersion("1.23.456-x.7.z.92");
+		testSemanticVersion("1.23.456-x.7.z.92+exp.sha.5114f85");
```
Comment (marcomarasca) [inline lib/lib-schema-id/src/test/java/org/sagebionetworks/schema/SchemaIdParserTest.java:170]: missing core+build only

File: lib/lib-schema-id/src/test/java/org/sagebionetworks/schema/SchemaIdParserTest.java [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+			testDotSeparatedAlphanumeric("1abc");
+		});
+	}
+	
+	@Test
+	public void testDotSeparatedAlphanumericStartsWithZero() {
+		assertThrows(ParseException.class, ()->{
+			testDotSeparatedAlphanumeric("0");
+		});
+	}
```
Comment (marcomarasca) [inline lib/lib-schema-id/src/test/java/org/sagebionetworks/schema/SchemaIdParserTest.java:204]: Add test for dotSeparatedAlphanumeritc that starts with a dot (and fails)

File: lib/lib-schema-id/src/test/java/org/sagebionetworks/schema/SchemaIdParserTest.java [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+	
+	@Test
+	public void testOrganziationName() throws ParseException {
+		testOrganizationName("foo");
+		testOrganizationName("foo.bar.a1");
+	}
+
```
Comment (marcomarasca) [inline lib/lib-schema-id/src/test/java/org/sagebionetworks/schema/SchemaIdParserTest.java:211]: Add tests for failing cases

File: lib/lib-schema-id/src/test/java/org/sagebionetworks/schema/SchemaIdParserTest.java [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+	public void testSchemaName() throws ParseException {
+		testSchemaName("SomeClass");
+		testSchemaName("repo.model.SomeClass");
+	}
+	
+	public void testSchemaName(String input) throws ParseException {
+		SchemaIdParser parser = new SchemaIdParser(input);
+		SchemaName schemaName = parser.schemaName();
+		assertEquals(input, schemaName.toString());
+	}
```
Comment (marcomarasca) [inline lib/lib-schema-id/src/test/java/org/sagebionetworks/schema/SchemaIdParserTest.java:228]: Add additional tests (failing/more coverage)

File: lib/lib-schema-id/src/test/java/org/sagebionetworks/schema/SchemaIdParserTest.java [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+	public void testSchemaIdWithVersion() throws ParseException {
+		SchemaId id = testSchemaId("org.myorg/path.SomeClass/1.2.3-alpha+1234f");
+		assertNotNull(id);
+		assertNotNull(id.getOrganizationName());
+		assertEquals("org.myorg",id.getOrganizationName().toString());
+		assertNotNull(id.getSchemaName());
+		assertEquals("path.SomeClass", id.getSchemaName().toString());
+		assertNotNull(id.getSemanticVersion());
+		assertEquals("1.2.3-alpha+1234f", id.getSemanticVersion().toString());
+	}
```
Comment (marcomarasca) [inline lib/lib-schema-id/src/test/java/org/sagebionetworks/schema/SchemaIdParserTest.java:251]: Add (failing) test for "org.myorg/path.SomeClass/1.2.03

File: lib/lib-schema-id/src/main/java/org/sagebionetworks/schema/id/OrganizationName.java [hunk 9]
```diff
@@ -0,0 +1,11 @@
+package org.sagebionetworks.schema.id;
+
+import org.sagebionetworks.schema.element.SimpleBranch;
+
+public class OrganizationName extends SimpleBranch {
```
Comment (marcomarasca) [inline lib/lib-schema-id/src/main/java/org/sagebionetworks/schema/id/OrganizationName.java:5]: Should extend DotSeparatedAlphaNumeric

File: lib/lib-schema-id/src/main/java/org/sagebionetworks/schema/id/SchemaId.java [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+	 */
+	public SemanticVersion getSemanticVersion() {
+		return semanticVersion;
+	}
+
+
+	@Override
+	public void toString(StringBuilder builder) {
+		organizationName.toString(builder);
+		builder.append("/");
```
Comment (marcomarasca) [inline lib/lib-schema-id/src/main/java/org/sagebionetworks/schema/id/SchemaId.java:56]: / can be constant

File: lib/lib-schema-id/src/main/java/org/sagebionetworks/schema/id/SchemaName.java [hunk 11]
```diff
@@ -0,0 +1,12 @@
+package org.sagebionetworks.schema.id;
+
+import org.sagebionetworks.schema.element.SimpleBranch;
+
+public class SchemaName extends SimpleBranch {
```
Comment (marcomarasca) [inline lib/lib-schema-id/src/main/java/org/sagebionetworks/schema/id/SchemaName.java:5]: Should extends DotSeparatedAlphaNumeric

File: pom.xml [hunk 12]
```diff
@@ -920,6 +926,16 @@
 					<artifactId>schema-to-pojo-maven-plugin</artifactId>
 					<version>${schema-to-pojo.version}</version>
 				</plugin>
+				<plugin>
+					<groupId>org.codehaus.mojo</groupId>
+					<artifactId>javacc-maven-plugin</artifactId>
+					<version>2.6</version>
```
Comment (marcomarasca) [inline pom.xml:932]: We can update?

File: pom.xml [hunk 13]
```diff
@@ -920,6 +926,16 @@
...
+				</plugin>
+				<plugin>
+					<groupId>org.codehaus.mojo</groupId>
+					<artifactId>build-helper-maven-plugin</artifactId>
+					<version>1.8</version>
```
Comment (marcomarasca) [inline pom.xml:937]: Can you add the same to the schema to pojo? :P

**Review by marcomarasca — APPROVED** [review]

File: lib/lib-schema-id/src/main/java/org/sagebionetworks/schema/id/OrganizationName.java (hunk 9 above)
Comment (john-hill) [reply lib/lib-schema-id/src/main/java/org/sagebionetworks/schema/id/OrganizationName.java:5]: I tried to do this but it is not compatible with how the parser works.

File: lib/lib-schema-id/src/main/java/org/sagebionetworks/schema/id/SchemaName.java (hunk 11 above)
Comment (john-hill) [reply lib/lib-schema-id/src/main/java/org/sagebionetworks/schema/id/SchemaName.java:5]: Same issue with OorganizaionName

File: pom.xml (hunk 13 above)
Comment (john-hill) [reply pom.xml:937]: It has already been added to schema-to-pojo-integration-tests pom.xml.  Have you run the build from the command line to generate the classes?

File: pom.xml (hunk 12 above)
Comment (john-hill) [reply pom.xml:932]: This is the latest version of the maven-plugin, even though there are newer versions of javacc.

**Pull request merged**